Apply featureFilterProvider in QgsVectorLayerDiagramProvider

arnaud-morvan opened the pull request:
## Description

Update of #8463 

As QgsVectorLayerDiagramProvider is created with it's "ownFeatureLoop", FILTER parameter was having no effect on diagrams.

Note: It should be possible to create QgsVectorLayerDiagramProvider with `ownFeatureLoop = true` here: https://github.com/qgis/QGIS/blob/9b04a29116156dec48c32dd6a54aafa0ff05eaea/src/core/qgsvectorlayerrenderer.cpp#L540

But this might have side effects.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

m-kuhn commented on src/core/qgsmaprenderercustompainterjob.cpp:
```diff
@@ -329,9 +329,15 @@ void QgsMapRendererJob::drawLabeling( const QgsMapSettings &settings, QgsRenderC
   // TODO: this is not ideal - we could override rendering stopped flag that has been set in meanwhile
   renderContext = QgsRenderContext::fromMapSettings( settings );
   renderContext.setPainter( painter );
+  renderContext.setFeatureFilterProvider( featureFilterProvider.get() );
```
The `featureFilterProvider` lives only in this method scope, but is set on the `renderContext` which lives outside this method.
Either pass ownership to the `renderContext` or reset the `featureFilterProvider` to its previous value before this method returns.

arnaud-morvan replied:
As far as I understand in renderContext keep a clone of the passed featureFilterProvider, see https://github.com/qgis/QGIS/blob/055a166136f29972571944bc741fbc68af42b4ea/src/core/qgsrendercontext.cpp#L244
This is why I want to delete temporary clone.

m-kuhn replied:
I see. So that's basically a double-clone?

Why not shorten it to

```
QgsFeatureFilterProvider *provider = renderContext.featureFilterProvider();
...
renderContext.setFeatureFilterProvider( featureFilterProvider );
```

arnaud-morvan replied:
This result in a SEGFAULT.
It seems that filter provider is destroyed after: 
`renderContext = QgsRenderContext::fromMapSettings( settings );`

Note that QgsRenderContext is a unique_ptr: https://github.com/qgis/QGIS/blob/a7c3aadf3d215774ee1202127718bdaa6437f71f/src/core/qgsrendercontext.h#L543

m-kuhn replied:
I'm still lost about the purpose of the renderContext in here. The existing context passed in is completely overwritten (which is probably not thread safe, as also indicated by the comment above `::fromMapSettings`) and this commit is the first one to introduce code that preserves anything from the previous renderContext.
I'm probably unaware of some background knowledge that led to this design decision, @wonder-sk do you remember the rationale for replacing the existing context? Should the featureFilterProvider be part of mapSettings?

m-kuhn replied:
If there are no further insights from @wonder-sk available, I agree it's best to stick to the proposed solution as anything else would require a more intense refactoring.

wonder-sk replied:
Hmm after looking into the code and history a bit, the whole reset of the render context in drawLabeling() seems useless. From what I understand, in older version of the code, this was the place where the actual render context for labeling got initialized. But since 33eb4bc0 from @nyalldawson there is a new prepareLabelingJob() method which does the same thing earlier (make context from map settings, set painter, extent, labeling engine). So if I didn't miss anything, the drawLabeling() body could be reduced to merely calling `labelingEngine2->run( renderContext );` and thus we would avoid this re-assignment of feature filter there (which indeed feels wrong)

m-kuhn commented on src/core/qgsmaprenderercustompainterjob.cpp:
```diff
@@ -329,9 +329,15 @@ void QgsMapRendererJob::drawLabeling( const QgsMapSettings &settings, QgsRenderC
   painter->setCompositionMode( QPainter::CompositionMode_SourceOver );
 
+  std::unique_ptr< QgsFeatureFilterProvider > featureFilterProvider;
+  if ( renderContext.featureFilterProvider() ) {
```
To make your life easier, setup the pre-commit hook ;)

arnaud-morvan committed:
Do not overwrite renderContext in QgsMapRendererJob::drawLabeling

Dedicated QgsRenderContext is already created and configured in prepareLabelingJob.

arnaud-morvan committed:
Apply featureFilterProvider in QgsVectorLayerDiagramProvider

arnaud-morvan commented:
Updated, it seems everything goes fine with passed QgsRenderContext.

m-kuhn commented on src/core/qgsvectorlayerdiagramprovider.cpp:
```diff
@@ -78,9 +78,13 @@ QList<QgsLabelFeature *> QgsVectorLayerDiagramProvider::labelFeatures( QgsRender
+  const QgsFeatureFilterProvider *featureFilterProvider = context.featureFilterProvider();
+  if ( featureFilterProvider )
+  {
+    featureFilterProvider->filterFeatures( qobject_cast<QgsVectorLayer *>( mLayer ), request );
```
Is it possible that this method is executed in a background thread?

In this case it's dangerous to use the layer directly and the feature source (`mSource`) should be used instead or bad things will happen (the application will crash in flames :boom: ) it looks to me as if this is an error by design in the feature filter provider.

nyalldawson replied:
It's an API design error, which was unfortunately only flagged after the feature filter work was merged (and we forgot to resolve for 3.0). It is very unsafe, but fortunately it's API which is only us by server for now (and maybe plugins?), and at least for server main thread behaviour is a bit more predictable.

The API needs fixing to remove this use of the layer pointer and replace it with something like the layer uri instead.

m-kuhn replied:
Anyway, it's not a blocker to this pull request, just something to keep in mind

arnaud-morvan commented:
For me everything is up to date.
I've created a new `drawLabeling` method without `settings` parameter and maked the old one as deprecated.

m-kuhn commented:
Thanks @arnaud-morvan
